man/show.man
.TH SHOW %manext1% 2016-03-24 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
show \- display nmh messages
.SH SYNOPSIS
.HP 5
.na
.B show
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msgs ]
.RB [ \-draft ]
.RB [ \-showproc
.IR program ]
.RB [ \-showmimeproc
.IR program ]
.RB [ \-header " | " \-noheader ]
.RB [ \-checkmime " | " \-nocheckmime ]
.RB [ \-concat " | " \-noconcat ]
[switches\ for
.I showproc
or
.IR showmimeproc ]
.ad
.SH DESCRIPTION
.B show
lists each of the specified messages to the standard output
(typically, the terminal).
.PP
By default, text (non-MIME) messages are filtered and displayed by
the
.B nmh
command
.BR mhl .
This command will display text
messages in a nice, uniform format.  It also allows you to configure
the format of the displayed messages and which headers fields are
shown.  See the
.IR mhl (1)
manual page for the details about this
command.  This default can be changed by defining the
.I showproc
profile component.  Any switches not recognized by
.BR show ,
as well as
.B \-header
and
.BR \-noheader ,
are
passed along to that program.  To override the default and the
.I showproc
profile component, use the
.B \-showproc
.I program
switch.  For example,
.B \-showproc
.I more
will cause the
.B more
program to list the messages with no reformatting.  Normally, this
program is specified as the
.I showproc
in the user's
.IR \&.mh_profile ,
rather than using a command line switch.
.PP
By default, non-text messages (MIME messages with multi-media
contents) are processed and displayed by the
.B nmh
command
.BR mhshow .
See the
.IR mhshow (1)
manual page for details
about this command.  This default can changed by defining the
.I showmimeproc
profile component.  Any switches not recognized
by
.B show
are passed along to that program.  To override this
default and the
.B showmimeproc
profile component, use the
.B \-showmimeproc
.I program
switch.
.PP
Note that in some cases,
.B show
may invoke the
.I showmimeproc
even for textual contents.  This will happen for text messages that
specify a transfer encoding (such as MIME quoted-printable or
base64) or specify a character set that
.B show
doesn't believe
can be displayed natively.  The appropriate
.IR locale (1)
environment variables
should be set to the terminal's native character set to avoid
gratuitous invocations of the
.IR showmimeproc .
See
.IR locale (1)
for details about these environment variables.
.PP
The option
.B \-checkmime
(set by default) instructs
.B show
to
test if any of the messages to be displayed are non-text (MIME)
messages.  If any are non-text, they are displayed by the program
.IR showmimeproc ,
else they are displayed by the program
.IR showproc .
The option
.B \-nocheckmime
disables this test and instructs
.B show
to use
.IR showproc ,
regardless of whether
any of the messages are non-text (MIME) messages.
.P
The
.B \-noshowproc
switch will disable any formatting or paging of
messages.  It is equivalent to
.B \-nocheckmime
.B \-showproc
.IR cat .
It is still accepted, but should be considered (somewhat) obsolete.
.PP
The
.B \-header
switch tells
.B show
to display a one-line
description of the message being shown.  This description includes
the folder and the message number.
.PP
By default
.B show
will concatenate all content under one pager.  If you want each part to
displayed separately, you can override the default behavior with
.B \-noconcat.
.PP
If no `msgs' are specified, the current message is used.  Although
it depends on the specific
.I showproc
or
.IR showmimeproc ,
in the default setup when more than one message is specified, you
will be prompted for a <RETURN> prior to listing each message.
Each message will be listed a page at a time, and when the end of
page is reached, the program will wait for a <SPACE> or <RETURN>.
If a <RETURN> is entered, it will print the next line, whereas
<SPACE> will print the next screenful.
.PP
If the standard output is not a terminal, no queries are made, and
each file is listed with a one-line header and two lines of
separation.
.PP
.RB \*(lq "show \-draft" \*(rq
will list the file <mh-dir>/draft if it
exists.
.PP
If the profile entry \*(lqUnseen\-Sequence\*(rq is present and
non-empty, then
.B show
will remove each of the messages shown
from each sequence named by the profile entry.
.SH FILES
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^$HOME/.mh_profile~^The user profile
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
^Path:~^To determine the user's nmh directory
^Current\-Folder:~^To find the default current folder
^Unseen\-Sequence:~^To name sequences denoting unseen messages
^showproc:~^Program to show text (non-MIME) messages
^showmimeproc:~^Program to show non-text (MIME) messages
.fi
.SH "SEE ALSO"
.IR mhl (1),
.IR mhshow (1),
.IR next (1),
.IR prev (1),
.IR scan (1)
.SH DEFAULTS
.nf
.RB ` +folder "' defaults to the current folder"
.RB ` msgs "' defaults to cur"
.RB ` \-checkmime '
.RB ` \-header '
.RB ` \-concat '
.fi
.SH CONTEXT
If a folder is given, it will become the current folder.  The last
message selected will become the current message.
.SH BUGS
The
.B \-header
switch doesn't work when `msgs' expands to more than
one message.  If the
.I showproc
is
.BR mhl ,
then this problem can
be circumvented by referencing the \*(lqmessagename\*(rq field in the
.B mhl
format file.
.PP
.B show
updates the user's context before showing the message.
Hence
.B show
will mark messages as seen prior to the user actually
seeing them.  This is generally not a problem, unless the user relies
on the \*(lqunseen\*(rq messages mechanism, and interrupts
.B show
while it is showing \*(lqunseen\*(rq messages.
.PP
If your
.I showproc
is
.B mhl
(the default), then
.B show
uses
a built-in
.BR mhl :
it does not actually run the
.B mhl
program.
Hence, if you define your own
.BR showproc ,
don't call it
.B mhl
since
.B show
won't run it.
.PP
If your
.I showproc
is the pager
.BR more ,
then avoid running
.B show
in the background with only its standard output piped to
another process, as in
.PP
.RS 5
show | imprint &
.RE
.PP
Due to a bug in
.BR more ,
show will go into a \*(lqtty input\*(rq state.
To avoid this problem, re-direct
.BR show 's
diagnostic output as well.
For users of
.BR csh :
.PP
.RS 5
show |& imprint &
.RE
.PP
For users of
.BR sh :
.PP
.RS 5
show 2>&1 | imprint &
.RE


NEWS
#
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.
- mhfixmsg now adds a Content-Transfer-Encoding header at the message level,
  if needed after decoding text parts.

uip/show.c
/* show.c -- show/list messages
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>
#include <h/mime.h>
#include <h/utils.h>
#include "sbr/m_maildir.h"

#define SHOW_SWITCHES \
    X("checkmime", 0, CHECKMIMESW) \
    X("nocheckmime", 0, NOCHECKMIMESW) \
    X("header", 0, HEADSW) \
    X("noheader", 0, NHEADSW) \
    X("form formfile", 0, FORMSW) \
    X("moreproc program", 0, PROGSW) \
    X("nomoreproc", 0, NPROGSW) \
    X("length lines", 0, LENSW) \
    X("width columns", 0, WIDTHSW) \
    X("showproc program", 0, SHOWSW) \
    X("showmimeproc program", 0, SHOWMIMESW) \
    X("noshowproc", 0, NSHOWSW) \
    X("draft", 0, DRFTSW) \
    X("file file", -4, FILESW) /* interface from showfile */ \
    X("fmtproc program", 0, FMTPROCSW) \
    X("nofmtproc", 0, NFMTPROCSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \
    /*				\
     * switches for mhlproc	\
     */				\
    X("concat", 0, CONCATSW) \
    X("noconcat", 0, NCONCATSW) \
    /*				\
     * switches for mhshow	\
     */				\
    X("part number", 0, PARTSW) \
    X("type content", 0, TYPESW) \
    X("prefer content", 0, PREFERSW) \
    X("markform file", 0, MARKFORMSW) \
    X("rcache policy", 0, RCACHESW) \
    X("wcache policy", 0, WCACHESW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(SHOW);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(SHOW, switches);
#undef X

/*
 * static prototypes
 */
static int is_nontext(char *);

#define	SHOW  0
#define	NEXT  1
#define	PREV  2


int
main (int argc, char **argv)
{
    int draftsw = 0, headersw = 1;
    int nshow = 0, checkmime = 1, mime = 0;
    int isdf = 0, mode = SHOW, msgnum;
    char *cp, *maildir, *file = NULL, *folder = NULL, *proc, *program;
    char buf[BUFSIZ], **argp, **arguments;
    struct msgs *mp = NULL;
    struct msgs_array msgs = { 0, 0, NULL };
    struct msgs_array vec = { 0, 0, NULL }, non_mhl_vec = { 0, 0, NULL };

    if (nmh_init(argv[0], 1)) { return 1; }

    if (!strcasecmp (invo_name, "next")) {
	mode = NEXT;
    } else if (!strcasecmp (invo_name, "prev")) {
	mode = PREV;
    }
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW:
		    ambigsw (cp, switches);
		    done (1);

		case HEADSW:
		    headersw = 1;
		    goto non_mhl_switches;
		case NHEADSW:
		    headersw = 0;
		    /* FALLTHRU */
		case CONCATSW:
		case NCONCATSW:
non_mhl_switches:
		    /* mhl can't handle these, so keep them separate. */
		    app_msgarg(&non_mhl_vec, --cp);
		    continue;

		case UNKWNSW:
		case NPROGSW:
		case NFMTPROCSW:
		    app_msgarg(&vec, --cp);
		    continue;

		case HELPSW:
		    snprintf (buf, sizeof(buf),
			"%s [+folder] %s[switches] [switches for showproc]",
			invo_name, mode == SHOW ? "[msgs] ": "");
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case DRFTSW:
		    if (file)
			adios (NULL, "only one file at a time!");
		    draftsw++;
		    if (mode == SHOW)
			continue;
usage:
		    adios (NULL,
			    "usage: %s [+folder] [switches] [switches for showproc]",
			    invo_name);
		case FILESW:
		    if (mode != SHOW)
			goto usage;
		    if (draftsw || file)
			adios (NULL, "only one file at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    file = path (cp, TFILE);
		    continue;

		case FORMSW:
		    app_msgarg(&vec, --cp);
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    app_msgarg(&vec, getcpy (etcpath(cp)));
		    continue;

		case PROGSW:
		case LENSW:
		case WIDTHSW:
		case FMTPROCSW:
		case PARTSW:
		case TYPESW:
		case PREFERSW:
		case MARKFORMSW:
		case RCACHESW:
		case WCACHESW:
		    app_msgarg(&vec, --cp);
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    app_msgarg(&vec, cp);
		    continue;

		case SHOWSW:
		    if (!(showproc = *argp++) || *showproc == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    nshow = 0;
		    continue;
		case NSHOWSW:
		    nshow++;
		    continue;

		case SHOWMIMESW:
		    if (!(showmimeproc = *argp++) || *showmimeproc == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    nshow = 0;
		    continue;
		case CHECKMIMESW:
		    checkmime++;
		    continue;
		case NOCHECKMIMESW:
		    checkmime = 0;
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
            folder = pluspath (cp);
	} else {
	    if (mode != SHOW)
		goto usage;
            app_msgarg(&msgs, cp);
	}
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    if (draftsw || file) {
	if (msgs.size)
	    adios (NULL, "only one file at a time!");
	if (draftsw)
	    app_msgarg(&vec, getcpy (m_draft (folder, NULL, 1, &isdf)));
	else
	    app_msgarg(&vec, file);
	headersw = 0;
	goto go_to_it;
    }

    if (!msgs.size) {
	switch (mode) {
	    case NEXT:
		app_msgarg(&msgs, "next");
		break;
	    case PREV:
		app_msgarg(&msgs, "prev");
		break;
	    default:
		app_msgarg(&msgs, "cur");
		break;
	}
    }

    if (!folder)
	folder = getfolder (1);
    maildir = m_maildir (folder);

    if (chdir (maildir) == NOTOK)
	adios (maildir, "unable to change directory to");

    /* read folder and create message structure */
    if (!(mp = folder_read (folder, 1)))
	adios (NULL, "unable to read folder %s", folder);

    /* check for empty folder */
    if (mp->nummsg == 0)
	adios (NULL, "no messages in %s", folder);

    /* parse all the message ranges/sequences and set SELECTED */
    for (msgnum = 0; msgnum < msgs.size; msgnum++)
	if (!m_convert (mp, msgs.msgs[msgnum]))
	    done (1);

    /*
     * Set the SELECT_UNSEEN bit for all the SELECTED messages,
     * since we will use that as a tag to know which messages
     * to remove from the "unseen" sequence.
     */
    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++)
	if (is_selected(mp, msgnum))
	    set_unseen (mp, msgnum);

    seq_setprev (mp);		/* set the Previous-Sequence */
    seq_setunseen (mp, 1);	/* unset the Unseen-Sequence */

    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++)
	if (is_selected(mp, msgnum))
	    app_msgarg(&vec, mh_xstrdup(m_name (msgnum)));

    seq_setcur (mp, mp->hghsel);	/* update current message  */
    seq_save (mp);			/* synchronize sequences   */
    context_replace (pfolder, folder);	/* update current folder   */
    context_save ();			/* save the context file   */

go_to_it: ;

    /*
     * Decide which "proc" to use
     */
    if (nshow) {
	proc = catproc;
    } else {
	/* check if any messages are non-text MIME messages */
	if (! mime  &&  checkmime) {
	    if (!draftsw && !file) {
		/* loop through selected messages and check for MIME */
		for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++)
		    if (is_selected (mp, msgnum) && is_nontext (m_name (msgnum))) {
			mime = 1;
			break;
		    }
	    } else {
		/* check the file or draft for MIME */
		if (is_nontext (vec.msgs[vec.size - 1]))
		    mime = 1;
	    }
	}

	/* Set the "proc" */
	if (mime)
	    proc = showmimeproc;
	else
	    proc = showproc;
    }

    if (folder && !draftsw && !file)
	setenv("mhfolder", folder, 1);

    if (strcmp (r1bindex (proc, '/'), "cat") == 0) {

	if (headersw && vec.size == 1)
	    printf ("(Message %s:%s)\n", folder, vec.msgs[0]);

    } else if (strcmp (r1bindex (proc, '/'), "mhl") == 0) {

	if (headersw && vec.size == 1)
	    printf ("(Message %s:%s)\n", folder, vec.msgs[0]);

	/* If "mhl", then run it internally */
	argsplit_insert(&vec, "mhl", &program);
	app_msgarg(&vec, NULL);
	mhl (vec.size, vec.msgs);
	done (0);

    } else {
	int i;
	char **mp;

	for (i = 0, mp = non_mhl_vec.msgs; i < non_mhl_vec.size; ++i, ++mp) {
	    if (draftsw || file) {
		/* Insert the switch before the filename. */
		app_msgarg(&vec, vec.msgs[vec.size - 1]);
		vec.msgs[vec.size - 2] = *mp;
	    } else {
		app_msgarg(&vec, *mp);
	    }
	}

	if (strcmp (r1bindex (proc, '/'), "mhn") == 0) {
	    /* Add "-file" if showing file or draft, */
	    if (draftsw || file) {
		app_msgarg(&vec, vec.msgs[vec.size - 1]);
		vec.msgs[vec.size - 2] = "-file";
	    }
	    /* and add -show for backward compatibility */
	    app_msgarg(&vec, "-show");
	} else if (strcmp (r1bindex (proc, '/'), "mhshow") == 0) {
	    /* If "mhshow", add "-file" if showing file or draft. */
	    if (draftsw || file) {
		app_msgarg(&vec, vec.msgs[vec.size - 1]);
		vec.msgs[vec.size - 2] = "-file";
	    }
	} else {
	    if (headersw && vec.size == 1)
		printf ("(Message %s:%s)\n", folder, vec.msgs[0]);
	}
    }

    fflush (stdout);

    argsplit_insert(&vec, proc, &program);
    app_msgarg(&vec, NULL);
    execvp (program, vec.msgs);
    adios (proc, "unable to exec");
    return 0;  /* dead code to satisfy the compiler */
}


/*
 * Check if a message or file contains any non-text parts
 */
static int
is_nontext (char *msgnam)
{
    int	result, state;
    char *bp, *dp, *cp;
    char buf[NMH_BUFSIZ], name[NAMESZ];
    FILE *fp;
    m_getfld_state_t gstate;

    if ((fp = fopen (msgnam, "r")) == NULL)
	return 0;
    gstate = m_getfld_state_init(fp);
    for (;;) {
	int bufsz = sizeof buf;
	switch (state = m_getfld2(&gstate, name, buf, &bufsz)) {
	case FLD:
	case FLDPLUS:
	    /*
	     * Check Content-Type field
	     */
	    if (!strcasecmp (name, TYPE_FIELD)) {
		int passno;
		char c;

		cp = mh_xstrdup(buf);
		while (state == FLDPLUS) {
		    bufsz = sizeof buf;
		    state = m_getfld2(&gstate, name, buf, &bufsz);
		    cp = add (buf, cp);
		}
		bp = cp;
		passno = 1;

again:
		for (; isspace ((unsigned char) *bp); bp++)
		    continue;
		if (*bp == '(') {
		    int i;

		    for (bp++, i = 0;;) {
			switch (*bp++) {
			case '\0':
invalid:
			    result = 0;
			    goto out;
			case '\\':
			    if (*bp++ == '\0')
				goto invalid;
			    continue;
			case '(':
			    i++;
			    continue;
			default:
			    continue;
			case ')':
			    if (--i < 0)
				break;
			    continue;
			}
			break;
		    }
		}
		if (passno == 2) {
		    if (*bp != '/')
			goto invalid;
		    bp++;
		    passno = 3;
		    goto again;
		}
		for (dp = bp; istoken (*dp); dp++)
		    continue;
		c = *dp;
		*dp = '\0';
		if (!*bp)
		    goto invalid;
		if (passno > 1) {
		    if ((result = (strcasecmp (bp, "plain") != 0)))
			goto out;
		    *dp = c;
		    for (dp++; isspace ((unsigned char) *dp); dp++)
			continue;
		    if (*dp) {
			if ((result = !uprf (dp, "charset")))
			    goto out;
			dp += LEN("charset");
			while (isspace ((unsigned char) *dp))
			    dp++;
			if (*dp++ != '=')
			    goto invalid;
			while (isspace ((unsigned char) *dp))
			    dp++;
			if (*dp == '"') {
			    if ((bp = strchr(++dp, '"')))
				*bp = '\0';
			} else {
			    for (bp = dp; *bp; bp++)
				if (!istoken (*bp)) {
				    *bp = '\0';
				    break;
				}
			}
		    } else {
			/* Default character set */
			dp = "US-ASCII";
		    }
		    /* Check the character set */
		    result = !check_charset (dp, strlen (dp));
		} else {
		    if (!(result = (strcasecmp (bp, "text") != 0))) {
			*dp = c;
			bp = dp;
			passno = 2;
			goto again;
		    }
		}
out:
		free (cp);
		if (result) {
		    fclose (fp);
		    m_getfld_state_destroy (&gstate);
		    return result;
		}
		break;
	    }

	    /*
	     * Check Content-Transfer-Encoding field
	     */
	    if (!strcasecmp (name, ENCODING_FIELD)) {
		cp = mh_xstrdup(buf);
		while (state == FLDPLUS) {
		    bufsz = sizeof buf;
		    state = m_getfld2(&gstate, name, buf, &bufsz);
		    cp = add (buf, cp);
		}
		for (bp = cp; isspace ((unsigned char) *bp); bp++)
		    continue;
		for (dp = bp; istoken ((unsigned char) *dp); dp++)
		    continue;
		*dp = '\0';
		result = (strcasecmp (bp, "7bit")
		       && strcasecmp (bp, "8bit")
		       && strcasecmp (bp, "binary"));

		free (cp);
		if (result) {
		    fclose (fp);
		    m_getfld_state_destroy (&gstate);
		    return result;
		}
		break;
	    }

	    /*
	     * Just skip the rest of this header
	     * field and go to next one.
	     */
	    while (state == FLDPLUS) {
		bufsz = sizeof buf;
		state = m_getfld2(&gstate, name, buf, &bufsz);
	    }
	    break;

	    /*
	     * We've passed the message header,
	     * so message is just text.
	     */
	default:
	    fclose (fp);
	    m_getfld_state_destroy (&gstate);
	    return 0;
	}
    }
}
